src/modules/airdrop/services/airdrop-service.ts

```diff
@@ -123,6 +132,91 @@ export class AirdropService {
     };
   }
 
+  async processMerkleDistributorRecipientsFile(file?: Express.Multer.File) {
+    if (!file) return;
+
+    const recipientsFile = await readFile(file.path, { encoding: "utf8" });
+    const recipientsJson = JSON.parse(recipientsFile);
+
+    try {
+      return await this.dataSource.transaction(async (entityManager) => {
+        const window = await entityManager
+          .createQueryBuilder()
+          .insert()
+          .into(MerkleDistributorWindow)
+          .values({
+            chainId: recipientsJson["chainId"],
+            rewardToken: recipientsJson["rewardToken"],
+            windowIndex: recipientsJson["windowIndex"],
+            rewardsToDeposit: recipientsJson["rewardsToDeposit"],
+            merkleRoot: recipientsJson["merkleRoot"],
+            ipfsHash: recipientsJson["ipfsHash"],
+          })
+          .execute();
+        const windowId = window.identifiers[0].id;
+
+        for (const recipient of Object.values(recipientsJson["recipientsWithProofs"])) {
+          await entityManager
+            .createQueryBuilder()
+            .insert()
+            .into(MerkleDistributorRecipient)
+            .values({
+              merkleDistributorWindowId: windowId,
+              address: recipient["account"],
+              amount: recipient["amount"],
+              accountIndex: recipient["accountIndex"],
+              proof: recipient["proof"],
+              payload: recipient["metadata"],
+            })
+            .execute();
+        }
+
+        const query = entityManager
+          .createQueryBuilder(MerkleDistributorRecipient, "recipient")
+          .select("COUNT(*) as count")
+          .where("recipient.merkleDistributorWindowId = :windowId", { windowId });
+        const recipientsCount = await query.getRawOne();
+
+        return {
+          recipients: isNaN(parseInt(recipientsCount.count)) ? null : parseInt(recipientsCount.count),
+        };
+      });
+    } catch (error) {
+      if (
+        error instanceof QueryFailedError &&
+        (error as QueryFailedError).driverError.constraint === "UK_merkle_distributor_window_windowIndex"
+      ) {
+        throw new DuplicatedMerkleDistributorWindowException();
+      }
+
+      throw error;
+    }
+  }
```

Comment:
I'm assuming this error just ends up causing the query to fail. Is that right?

Outside of the query failing, do we have any way of logging or reporting errors, especially bugs/internal issues (as opposed to user input issues)?

Reply:
@mrice32 
Any error caused by a failed query will make the transaction to revert. The reason I want to handle this error separately is the fact that any thrown NodeJS error will trigger a 500 Internal Server HTTP error, but for duplicated uploaded MerkleDistributor windows I want the backend to respond with a custom error (custom message and custom HTTP code). Custom HTTP errors can be created by instantiating custom HTTPException errors

> do we have any way of logging or reporting errors
For now we only log errors inside the container. Any thrown error can be seen there